_posts/2022-04-01-First-steps-with-swift-concurrency.md

```diff
@@ -0,0 +1,341 @@
+
+In addition, when there is a new thumbnail fetched, we want to inform UI about it to refresh the view and present the loaded images on screen.
+
+The best candidates to fulfill those requirements are:
+
+- `Task Group` which allows us to create mutliple concurrent tasks for fetching the thumbnails.
+- `AsyncStream` which allows us to receive the data when they are delivered one after another.
+
+So first of all - we start with creating the new data type which will perform all of this. Once again `actor` is a great choice since all of those thing are not user-facing, so it should be done in background. Let's call it `ThumbnailDataSource`.
+
+```swift
+public actor ThumbnailDataSource { }
+```
+
+Great, now we define the function which will start fetching the thumbnails.
+
+```swift
+public actor ThumbnailDataSource { 
+	public typealias ThumbnailsStream = AsyncStream<(String, UIImage)>
+
+	public nonisolated func getThumbnails(models: [ApodModel]) -> ThumbnailsStream {
+		return ThumbnailsStream { continuation in
+			Task {
+				try await fetchThumbnails(from: models, continuation: continuation)
+			}
+		}
+	}
+}
+```
+
+As a parameter we take the array of ApodModels which contains the url for our images. Our function will return the `AsyncStream`, which will be delivering the tuple `(String, UIImage)` when the particular image is fetched (`String` will contain the image URL).
+
+Now we create the `AsyncStream` using its intializer, which is returning the `continuation` as a closure parameter, which we will use later on (more about [AsyncStream](https://developer.apple.com/documentation/swift/asyncstream)). In short continuation is needed to inform the stream about the received elements or when the stream is finishing its work.
+
+Please note that `func getThumbnails()` is not an `async` function, because we would like to only bind the data and receive them later on. That's why all async code was wrapped as a `Task`. The `nonisolated` prefix gave a possibility for the function to synchronously create the `AsyncStream` in the code. Since we do not modify anything in terms of the actors properties in the function, it is safe to expose the method as `nonisolated`.
+
+And then the usage of the Task Group looks like below:
+
+```swift
+private func fetchThumbnails(
+        from apody: [ApodModel],
+        continuation: ThumbnailsStream.Continuation
+    ) async throws {
+        try await withThrowingTaskGroup(of: (String, UIImage).self) { group in
+            for apod in apody.filter({ $0.media_type == .image }) {
+                group.addTask(priority: .background) {
+                    (apod.url, try await self.fetchImage(url: apod.url))
+                }
+            }
```

Comment:
```suggestion
        from apods: [ApodModel],
        continuation: ThumbnailsStream.Continuation
    ) async throws {
        try await withThrowingTaskGroup(of: (String, UIImage).self) { group in
            for apod in apods.filter({ $0.media_type == .image }) {
                group.addTask(priority: .background) {
                    (apod.url, try await self.fetchImage(url: apod.url))
                }
            }
```

Was `apody` a plural in Polish :D ?

Reply:
Haha, good one 😅 Improved every single "apods" phrase 😄